quarkchain/cluster/cluster_config.py

```diff
@@ -91,17 +91,49 @@ class SlaveConfig(BaseConfig):
     PORT = 38392
     WEBSOCKET_JSON_RPC_PORT = None
     ID = ""
-    CHAIN_MASK_LIST = None
+    FULL_SHARD_ID_LIST = []
+
+    # private configurable variable to handle legacy CHAIN_MASK_LIST translation
+    _chains = None
 
     def to_dict(self):
         ret = super().to_dict()
-        ret["CHAIN_MASK_LIST"] = [m.value for m in self.CHAIN_MASK_LIST]
+        # format to hex
+        ret["FULL_SHARD_ID_LIST"] = [
+            "0x{:08x}".format(i) for i in self.FULL_SHARD_ID_LIST
+        ]
         return ret
 
     @classmethod
     def from_dict(cls, d):
         config = super().from_dict(d)
-        config.CHAIN_MASK_LIST = [ChainMask(v) for v in config.CHAIN_MASK_LIST]
+        # bail if both full shard ID list and chain mask list exist
+        shard_ids = getattr(config, "FULL_SHARD_ID_LIST", None)
+        chain_mask = getattr(config, "CHAIN_MASK_LIST", None)
+        if shard_ids and chain_mask:
+            raise ValueError(
+                "Can only have either FULL_SHARD_ID_LIST or CHAIN_MASK_LIST"
+            )
+        elif shard_ids:
+            # parse from hex to int
+            config.FULL_SHARD_ID_LIST = [int(h, 16) for h in config.FULL_SHARD_ID_LIST]
+        elif chain_mask:
+            # a simple way to be backward compatible with hard-coded shard ID
+            # e.g. chain mask 4 => 0x00000001, 0x00040001
+            # note that this only works if every chain has 1 shard only
+            check(all(chain.SHARD_SIZE == 1 for chain in cls._chains))
```

Comment:
could we have an optional argument with chains instead of a global class variable?  It seems that "chains" is known to the caller upon calling from_dict.  This also minimizes the scope of the "chains" variable.

Reply:
done